# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

-   Acquire data in various ways
-   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="#data_acquisition">Data Acquisition</a>
    <li><a href="#basic_insight">Basic Insight of Dataset</a></li>
</ol>


</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Automobile Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.
<ul>
    <li>Data info: <a href="https://archive.ics.uci.edu/dataset/10/automobile" >https://archive.ics.uci.edu/dataset/10/automobile</a></li>
    <li>Data source: <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork971-2022-01-01" target="_blank">https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


In [45]:
# import pandas library
import pandas as pd
import numpy as np

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

Because the data does not include headers, we can add an argument <code>headers = None</code> inside the <code>read_csv()</code> method so that pandas will not automatically set the first row as a header.<br>

You can also assign the dataset to any variable you create.
</p>


In [46]:
# Read the online file by the URL provides above, and assign it to variable "df"
# YOUR CODE HERE
df = pd.read_csv("C:/Users/tontide1/Desktop/Data/Programing and statistics data/labs-20240821T054947Z-001/labs/auto.csv")

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [47]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...  \
0  3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1  1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
2  2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
3  2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   
4  2    ?         audi  gas  std   two        sedan  fwd  front  99.8  ...   

   130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
1  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
2  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
3  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
4  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  

[5 rows x 26 columns]

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [48]:
# YOUR CODE HERE
df.tail(10)

3    ? alfa-romero     gas    std   two convertible  rwd  front   88.6  \
194 -1   74       volvo     gas    std  four       wagon  rwd  front  104.3   
195 -2  103       volvo     gas    std  four       sedan  rwd  front  104.3   
196 -1   74       volvo     gas    std  four       wagon  rwd  front  104.3   
197 -2  103       volvo     gas  turbo  four       sedan  rwd  front  104.3   
198 -1   74       volvo     gas  turbo  four       wagon  rwd  front  104.3   
199 -1   95       volvo     gas    std  four       sedan  rwd  front  109.1   
200 -1   95       volvo     gas  turbo  four       sedan  rwd  front  109.1   
201 -1   95       volvo     gas    std  four       sedan  rwd  front  109.1   
202 -1   95       volvo  diesel  turbo  four       sedan  rwd  front  109.1   
203 -1   95       volvo     gas  turbo  four       sedan  rwd  front  109.1   

     ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
194  ...  141  mpfi  3.78  3.15   9.5  114  5400  23  28  13415  
195  ...  141  mpfi  3.78  3.15   9.5  114  5400  24  28  15985  
196  ...  141  mpfi  3.78  3.15   9.5  114  5400  24  28  16515  
197  ...  130  mpfi  3.62  3.15   7.5  162  5100  17  22  18420  
198  ...  130  mpfi  3.62  3.15   7.5  162  5100  17  22  18950  
199  ...  141  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
200  ...  141  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
201  ...  173  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
202  ...  145   idi  3.01  3.40  23.0  106  4800  26  27  22470  
203  ...  141  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[10 rows x 26 columns]

<h3>Add Headers</h3>
<p>
Take a look at our dataset. Pandas automatically set the header with an integer starting from 0.
</p>
<p>
To better describe our data, we can introduce a header. This information is available at:  <a href="https://archive.ics.uci.edu/ml/datasets/Automobile?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork971-2022-01-01" target="_blank">https://archive.ics.uci.edu/ml/datasets/Automobile</a>.
</p>
<p>
Thus, we have to add headers manually.
</p>
<p>
First, we create a list "headers" that include all column names in order.
Then, we use <code>dataframe.columns = headers</code> to replace the headers with the list we created.
</p>


In [49]:
# create headers list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


 We replace headers and recheck our dataframe:


In [50]:
# YOUR CODE HERE
df.columns = headers
df

symboling normalized-losses         make fuel-type aspiration  \
0            3                 ?  alfa-romero       gas        std   
1            1                 ?  alfa-romero       gas        std   
2            2               164         audi       gas        std   
3            2               164         audi       gas        std   
4            2                 ?         audi       gas        std   
..         ...               ...          ...       ...        ...   
199         -1                95        volvo       gas        std   
200         -1                95        volvo       gas      turbo   
201         -1                95        volvo       gas        std   
202         -1                95        volvo    diesel      turbo   
203         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two    hatchback          rwd           front        94.5  ...   
2           four        sedan          fwd           front        99.8  ...   
3           four        sedan          4wd           front        99.4  ...   
4            two        sedan          fwd           front        99.8  ...   
..           ...          ...          ...             ...         ...  ...   
199         four        sedan          rwd           front       109.1  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            152         mpfi  2.68    3.47               9.0        154   
2            109         mpfi  3.19    3.40              10.0        102   
3            136         mpfi  3.19    3.40               8.0        115   
4            136         mpfi  3.19    3.40               8.5        110   
..           ...          ...   ...     ...               ...        ...   
199          141         mpfi  3.78    3.15               9.5        114   
200          141         mpfi  3.78    3.15               8.7        160   
201          173         mpfi  3.58    2.87               8.8        134   
202          145          idi  3.01    3.40              23.0        106   
203          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  16500  
1        5000       19          26  16500  
2        5500       24          30  13950  
3        5500       18          22  17450  
4        5500       19          25  15250  
..        ...      ...         ...    ...  
199      5400       23          28  16845  
200      5300       19          25  19045  
201      5500       18          23  21485  
202      4800       26          27  22470  
203      5400       19          25  22625  

[204 rows x 26 columns]

We need to replace the "?" symbol with NaN so the dropna() can remove the missing values:


In [51]:
# YOUR CODE HERE
df.replace("?", np.nan, inplace=True)
df.head(20)

symboling normalized-losses         make fuel-type aspiration  \
0           3               NaN  alfa-romero       gas        std   
1           1               NaN  alfa-romero       gas        std   
2           2               164         audi       gas        std   
3           2               164         audi       gas        std   
4           2               NaN         audi       gas        std   
5           1               158         audi       gas        std   
6           1               NaN         audi       gas        std   
7           1               158         audi       gas      turbo   
8           0               NaN         audi       gas      turbo   
9           2               192          bmw       gas        std   
10          0               192          bmw       gas        std   
11          0               188          bmw       gas        std   
12          0               188          bmw       gas        std   
13          1               NaN          bmw       gas        std   
14          0               NaN          bmw       gas        std   
15          0               NaN          bmw       gas        std   
16          0               NaN          bmw       gas        std   
17          2               121    chevrolet       gas        std   
18          1                98    chevrolet       gas        std   
19          0                81    chevrolet       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two    hatchback          rwd           front        94.5  ...   
2          four        sedan          fwd           front        99.8  ...   
3          four        sedan          4wd           front        99.4  ...   
4           two        sedan          fwd           front        99.8  ...   
5          four        sedan          fwd           front       105.8  ...   
6          four        wagon          fwd           front       105.8  ...   
7          four        sedan          fwd           front       105.8  ...   
8           two    hatchback          4wd           front        99.5  ...   
9           two        sedan          rwd           front       101.2  ...   
10         four        sedan          rwd           front       101.2  ...   
11          two        sedan          rwd           front       101.2  ...   
12         four        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       103.5  ...   
14         four        sedan          rwd           front       103.5  ...   
15          two        sedan          rwd           front       103.5  ...   
16         four        sedan          rwd           front       110.0  ...   
17          two    hatchback          fwd           front        88.4  ...   
18          two    hatchback          fwd           front        94.5  ...   
19         four        sedan          fwd           front        94.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           152         mpfi  2.68    3.47               9.0        154   
2           109         mpfi  3.19    3.40              10.0        102   
3           136         mpfi  3.19    3.40               8.0        115   
4           136         mpfi  3.19    3.40               8.5        110   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           131         mpfi  3.13    3.40               8.3        140   
8           131         mpfi  3.13    3.40               7.0        160   
9           108         mpfi  3.50    2.80               8.8        101   
10          108         mpfi  3.50    2.80               8.8        101   
11          164

We can drop missing values along the column "price" as follows:


In [52]:
# YOUR CODE HERE
df = df[df['price'].notna()]
df.head(20)

symboling normalized-losses         make fuel-type aspiration  \
0           3               NaN  alfa-romero       gas        std   
1           1               NaN  alfa-romero       gas        std   
2           2               164         audi       gas        std   
3           2               164         audi       gas        std   
4           2               NaN         audi       gas        std   
5           1               158         audi       gas        std   
6           1               NaN         audi       gas        std   
7           1               158         audi       gas      turbo   
9           2               192          bmw       gas        std   
10          0               192          bmw       gas        std   
11          0               188          bmw       gas        std   
12          0               188          bmw       gas        std   
13          1               NaN          bmw       gas        std   
14          0               NaN          bmw       gas        std   
15          0               NaN          bmw       gas        std   
16          0               NaN          bmw       gas        std   
17          2               121    chevrolet       gas        std   
18          1                98    chevrolet       gas        std   
19          0                81    chevrolet       gas        std   
20          1               118        dodge       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two    hatchback          rwd           front        94.5  ...   
2          four        sedan          fwd           front        99.8  ...   
3          four        sedan          4wd           front        99.4  ...   
4           two        sedan          fwd           front        99.8  ...   
5          four        sedan          fwd           front       105.8  ...   
6          four        wagon          fwd           front       105.8  ...   
7          four        sedan          fwd           front       105.8  ...   
9           two        sedan          rwd           front       101.2  ...   
10         four        sedan          rwd           front       101.2  ...   
11          two        sedan          rwd           front       101.2  ...   
12         four        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       103.5  ...   
14         four        sedan          rwd           front       103.5  ...   
15          two        sedan          rwd           front       103.5  ...   
16         four        sedan          rwd           front       110.0  ...   
17          two    hatchback          fwd           front        88.4  ...   
18          two    hatchback          fwd           front        94.5  ...   
19         four        sedan          fwd           front        94.5  ...   
20          two    hatchback          fwd           front        93.7  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68              9.00        111   
1           152         mpfi  2.68    3.47              9.00        154   
2           109         mpfi  3.19    3.40             10.00        102   
3           136         mpfi  3.19    3.40              8.00        115   
4           136         mpfi  3.19    3.40              8.50        110   
5           136         mpfi  3.19    3.40              8.50        110   
6           136         mpfi  3.19    3.40              8.50        110   
7           131         mpfi  3.13    3.40              8.30        140   
9           108         mpfi  3.50    2.80              8.80        101   
10          108         mpfi  3.50    2.80              8.80        101   
11          164         mpfi  3.31    3.19              9.00        121   
12          164

Now, we have successfully read the raw dataset and added the correct headers into the dataframe.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [53]:
# Write your code below and press Shift+Enter to execute 
# YOUR CODE HERE
print(df.columns.values)


['symboling' 'normalized-losses' 'make' 'fuel-type' 'aspiration'
 'num-of-doors' 'body-style' 'drive-wheels' 'engine-location' 'wheel-base'
 'length' 'width' 'height' 'curb-weight' 'engine-type' 'num-of-cylinders'
 'engine-size' 'fuel-system' 'bore' 'stroke' 'compression-ratio'
 'horsepower' 'peak-rpm' 'city-mpg' 'highway-mpg' 'price']


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>automobile.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


In [55]:
# YOUR CODE HERE
df.to_csv("auto_new.csv", na_rep="NaN")

We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.
</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:
</p>


In [56]:
df.dtypes


symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

A series with the data type of each column is returned.


In [57]:
# check the data type of data frame "df" by .dtypes
# YOUR CODE HERE
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

<p>
As shown above, it is clear to see that the data type of "symboling" and "curb-weight" are <code>int64</code>, "normalized-losses" is <code>object</code>, and "wheel-base" is <code>float64</code>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [58]:
df.describe()

symboling  wheel-base      length       width      height  \
count  200.000000  200.000000  200.000000  200.000000  200.000000   
mean     0.830000   98.848000  174.228000   65.898000   53.791500   
std      1.248557    6.038261   12.347132    2.102904    2.428449   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.675000   64.175000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.500000   66.675000   55.525000   
max      3.000000  120.900000  208.100000   72.000000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   200.000000   200.000000         200.000000  200.000000   200.000000  
mean   2555.705000   126.860000          10.170100   25.200000    30.705000  
std     518.594552    41.650501           4.014163    6.432487     6.827227  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2163.000000    97.750000           8.575000   19.000000    25.000000  
50%    2414.000000   119.500000           9.000000   24.000000    30.000000  
75%    2928.250000   142.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.
<br>

However, what if we would also like to check all the columns including those that are of type object?
<br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [59]:
# describe all the columns in "df" 
# YOUR CODE HERE
df.describe(include="all")

symboling normalized-losses    make fuel-type aspiration  \
count   200.000000               164     200       200        200   
unique         NaN                51      22         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      32       180        164   
mean      0.830000               NaN     NaN       NaN        NaN   
std       1.248557               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           198        200          200             200  200.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            113         94          118             197         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.848000  ...   
std             NaN        NaN          NaN             NaN    6.038261  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    200.000000          200   196     196        200.000000        198   
unique          NaN            8    38      36               NaN         58   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           91    23      19               NaN         19   
mean     126.860000          NaN   NaN     NaN         10.170100        NaN   
std       41.650501          NaN   NaN     NaN          4.014163        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.750000          NaN   NaN     NaN          8.575000        NaN   
50%      119.500000          NaN   NaN     NaN          9.000000        NaN   
75%      142.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg  price  
count        198  200.000000  200.000000    200  
unique        22         NaN         NaN    185  
top         5500         NaN         NaN  16500  
freq          36         NaN         NaN      2  
mean         NaN   25.200000   30.705000    NaN  
std          NaN    6.432487    6.827227    NaN  
min          NaN   13.000000   16.000000    NaN  
25%          NaN   19.000000   25.000000    NaN  
50%          NaN   24.000000   30.000000    NaN  
75%          NaN   30.000000   34.000000    NaN  
max          NaN   49.000000   54.000000    NaN  

[11 rows x 26 columns]

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>
</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'length' and 'compression-ratio'.

</div>


In [60]:
# YOUR CODE HERE
df[['length', 'compression-ratio']].describe()

length  compression-ratio
count  200.000000         200.000000
mean   174.228000          10.170100
std     12.347132           4.014163
min    141.100000           7.000000
25%    166.675000           8.575000
50%    173.200000           9.000000
75%    183.500000           9.400000
max    208.100000          23.000000

<h2>Info</h2>
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame. 

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [61]:
# look at the info of "df"
# YOUR CODE HERE
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 200 entries, 0 to 203
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          200 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               200 non-null    object 
 3   fuel-type          200 non-null    object 
 4   aspiration         200 non-null    object 
 5   num-of-doors       198 non-null    object 
 6   body-style         200 non-null    object 
 7   drive-wheels       200 non-null    object 
 8   engine-location    200 non-null    object 
 9   wheel-base         200 non-null    float64
 10  length             200 non-null    float64
 11  width              200 non-null    float64
 12  height             200 non-null    float64
 13  curb-weight        200 non-null    int64  
 14  engine-type        200 non-null    object 
 15  num-of-cylinders   200 non-null    object 
 16  engine-size        200 non-null

---